src/retry-queue.js

```diff
@@ -0,0 +1,108 @@
+import { RequestQueueScaffold } from './base-request-queue'
+import { encodePostData, xhr } from './send-request'
+
+export class RetryQueue extends RequestQueueScaffold {
+    constructor(captureMetrics) {
+        super()
+        this.captureMetrics = captureMetrics
+        this.isPolling = false
+        this._requestRetriesMap = {} // <RequestId, number>
+        this._counterToQueueMap = {}
+        this._pollerCounter = 1
+        this._areWeOnline = 'onLine' in window.navigator ? window.navigator.onLine : true
+        this._offlineBacklog = []
+
+        if ('onLine' in window.navigator) {
+            window.addEventListener('online', () => {
+                this._handleWeAreNowOnline()
+            })
+            window.addEventListener('offline', () => {
+                this._areWeOnline = false
+            })
+        }
+    }
+
+    enqueue(requestData) {
+        const { requestId } = requestData
+        let retriesPerformedSoFar = 0
+        if (!(requestId in this._requestRetriesMap)) {
+            this._requestRetriesMap[requestId] = 0
+        } else {
+            retriesPerformedSoFar = this._requestRetriesMap[requestId]
+            if (retriesPerformedSoFar === 10) {
+                delete this._requestRetriesMap[requestId]
+                return
+            }
+            this._requestRetriesMap[requestId]++
+        }
+
+        const nextRetry = this._pollerCounter - 1 + 2 ** retriesPerformedSoFar
+        if (!(nextRetry in this._counterToQueueMap)) {
+            this._counterToQueueMap[nextRetry] = []
+        }
+        this._counterToQueueMap[nextRetry].push(requestData)
```

Comment:
This looks very complicated :D. 
I see how it works in the end, but I had to really dig into the code to understand what's going on. 

Could we replace this step counter system with a simple timestamp-based key/value store? Entries are enqueued by timestamp, and all in the past get run each poll step.